src/audio/eq_iir/llext/CMakeLists.txt

```diff
@@ -0,0 +1,53 @@
+# Copyright (c) 2024 Intel Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_path(SET SOF_BASE NORMALIZE ${CMAKE_CURRENT_SOURCE_DIR}/../../../..)
+
+set(MODULE "eq_iir")
+
+file(STRINGS ${CMAKE_CURRENT_LIST_DIR}/../${MODULE}.toml uuids REGEX "^[ \t]*uuid *=")
+
+foreach(line IN LISTS uuids)
+	# extract UUID value - drop the 'uuid = ' part of the assignment line
+	string(REGEX REPLACE "^[ \t]*uuid *= \"([0-9A-F\-]*)\"" "\\1" uuid ${line})
+	file(APPEND ${ZEPHYR_BINARY_DIR}/${MODULE}_llext/llext.uuid "${uuid}\n")
+endforeach()
+
+add_llext_target(${MODULE}
+	OUTPUT  ${PROJECT_BINARY_DIR}/${MODULE}_llext/${MODULE}.llext
+	SOURCES ../eq_iir.c
+		../eq_iir_ipc4.c
+		../eq_iir_generic.c
+)
+
+target_include_directories(${MODULE}_llext_lib PRIVATE
+	"${SOF_BASE}/xtos/include"
+	"${SOF_BASE}/src/include"
+	"${SOF_BASE}/tools/rimage/src/include"
+)
+
+sof_append_relative_path_definitions(${MODULE}_llext_lib)
+
+add_llext_command(TARGET ${MODULE}
+	PRE_BUILD
+	COMMAND ${CMAKE_C_COMPILER} -E ${CMAKE_CURRENT_LIST_DIR}/llext.toml.h -P -DREM=
+		-I${SOF_BASE} -I${SOF_BASE}src
+		-imacros ../include/generated/autoconf.h
+		-o rimage_config.toml
+)
+
+if("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "zephyr")
+set(EXTRA_LINKER_PARAMS -nostdlib -nodefaultlibs -shared)
+else()
+set(EXTRA_LINKER_PARAMS -nostdlib -nodefaultlibs -r)
+endif()
+
+get_target_property(proc_in_file ${MODULE} lib_output)
+get_target_property(proc_out_file ${MODULE} pkg_input)
+add_llext_command(TARGET ${MODULE}
+	POST_BUILD
+	COMMAND ${SOF_BASE}scripts/llext_link_helper.py
+		--text-addr="0xa06ea000" -f ${proc_in_file} ${CMAKE_C_COMPILER} --
```

Comment:
where does this address come from ? Does it need to come from platform / DT or Kconfig ?

Reply:
@lgirdwood right, now this is a good question, let's talk about this. These address values come originally from the original loadable module concept, whereby we build modules for a fixed address in memory - I think that holds for system services too, probably for IADK as well. Having those addresses helps - it makes linking at run time easier, internal address references are already correct, but it does mean that we need to maintain those addresses somehow somewhere. The original one, that is now used for the smart-amp was exactly the address used in some old examples. Now for mixin-mixout and eq-iir I picked up some addresses "similar" to smart-amp but far enough from it... But you're right, we need to come up with a system of how to assign these addresses or maybe to aim to get rid of them completely and to perform full run-time linking / relocation instead

Reply:
At the very least don't copy/paste/diverge it in every module. Define it in one place and ta-da: you can paste your explanation above in that place.

Reply:
yeah, we can have a comment today and have a per platform memory address for this in the Zephyr llext core as next steps, probably best to align with @teburd

Reply:
I had some ideas around this by reusing the devicetree flash partitions or something like it, which is also used by mcuboot to designate memory regions for images.

E.g. imagine a set of partitions for fixed location modules/applications akin to this dts:
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/frdm_k64f/frdm_k64f.dts#L229

I think @dapperlo would also be interested (also using parts of llext) https://github.com/zephyrproject-rtos/zephyr/pull/72258

Reply:
Interested indeed. I need to ramp on the context for this use-case, but there is a desire for fixed address extensions and the load-time benefits that comes from it.

Nodes in the dts sounds reasonable, routing loaded `.llext` blobs could then use a label/attribute in blob header to associate.

Reply:
> yeah, we can have a comment today and have a per platform memory address for this in the Zephyr llext core as next steps, probably best to align with @teburd

It isn't just per platform, it's also per module. This is the starting address of *this module* on *this platform*. So, we could have a per-platform address range for loadable modules and then per-module offsets defined in modules themselves? But that would mean some co-ordination in terms of sufficiently large offsets. I'm not sure how large is the usable address range e.g. on MTL? If it's large enough we could just dedicate an MB per module?

Alternatively: we already have TOMLs per platform that include all modules, so I'm wondering if we could add in all those locations like https://github.com/thesofproject/sof/blob/374d2d668d97acb78c7e08c08da7a898d542b2c1/tools/rimage/config/mtl.toml.h#L29 `#define ${MODULE}_base 0x...` statements (e.g. `#define smart_amp_test_base 0xa06ca000`) but I'm not sure we'd be able to pull those out into cmake.

Reply:
> E.g. imagine a set of partitions described for fixed location modules/applications akin to this dts:
> https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/frdm_k64f/frdm_k64f.dts#L229

@teburd yes, looks good, with a slight inconvenience that we now have on MTL something like 28 modules. Maybe a separate .dts, included into the main one

Reply:
I have no issue with toml either but I’ll need help in understanding how we piece this together. Do we need to link the extensions to be placed appropriately with a custom linker script per extension slot? Sure seems like it to me at first glance but please do correct me if wrong.

Reply:
@teburd no, I tried very hard and so far have succeeded in avoiding custom linker scripts. I "only" provide section addresses to the linker on the command line. For that I only take a single base address as input and calculate all further addresses automatically.